### PR TITLE
CSS: Avoid filling jQuery.cssProps

### PR DESCRIPTION
**Pull request by dmethvin**
Fixes gh-3986

This avoids filling jQuery.cssProps by introducing a second internal
prop cache. It allows jQuery Migrate to detect external usage.

* [x] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* [ ] ~~If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com~~

### CODE REVIEW

**Comment by mgol** [comment]
What's the size difference?

File: src/css/finalPropName.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+define( [ "../var/document" ], function( document ) {
+
+"use strict";
```
Comment (mgol) [inline src/css/finalPropName.js]: Please add an empty line below this one; I think that's what we do elsewhere.

File: src/effects/Tween.js [hunk 2]
```diff
@@ -14,6 +16,8 @@ Tween.prototype = {
 	constructor: Tween,
 	init: function( elem, options, prop, end, easing, unit ) {
 		this.elem = elem;
+		this.styleElem = elem.nodeType === 1 && ( jQuery.cssHooks[ prop ] ||
```
Comment (mgol) [inline src/effects/Tween.js]: What's the reason for this change? We're now unit-testing `jQuery.Tween`, in particular the behavior of its properties so it's kind of a public API now. I'd rather not expose internal properties on Tween objects.

**Comment by mgol** [comment]
Unit tests are failing in Travis.

File: src/effects/Tween.js (hunk 2 above)
Comment (dmethvin) [reply src/effects/Tween.js]: Mainly I was trying to avoid doing this calculation each time and just do it once in the init. It's no problem to move it back though.

**Comment by dmethvin** [comment]
Not sure why I didn't see those failures. Moving the test back out of the init fixed the issue, I guess if all props in the Tween are considered fair game at any time that's not a good optimization.

 I did notice that in the tween.js tests we're mocking `jQuery.style()` which isn't a public interface.

**Comment by dmethvin** [comment]
@mgol was there something else you wanted to review before I landed this? I noticed the tag.

File: src/css/finalPropName.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		name = cssPrefixes[ i ] + capName;
+		if ( name in emptyStyle ) {
+			return name;
+		}
+	}
+}
+
+// Return a potentially-mapped jQuery.cssProps or vendor prefixed property
+function finalPropName( name ) {
+	var final = jQuery.cssProps[ name ] || vendorProps[ name ];
```
Comment (Krinkle) [inline src/css/finalPropName.js]: Hm.. I don't think this resolved the concern from https://github.com/jquery/jquery-migrate/pull/291.

The jquery-migrate pull proposes a warning for getting (and theoretically, setting) of `jQuery.cssProps` itself. That may seem odd, but is actually fine, given that detecting the setting of sub-properties is hard, and, in order to set a sub-property, one has to get the cssProps object first, which will trigger the warning. All good.

However, the code here in core/css.js, still gets the `jQuery.cssProps` object the same way as before, so it seems that would trigger the warning?

The only thing that changed is it now gets the object once instead of twice.

Comment (dmethvin) [reply src/css/finalPropName.js]: In the previous code, we **set** a property on `jQuery.cssProps`. Now we no longer do that, so Migrate can look for situations where external code sets values here.

Comment (Krinkle) [reply src/css/finalPropName.js]: Yep, but the warning added in https://github.com/jquery/jquery-migrate/pull/291 is not for getting/setting unknown properties on `jQuery.cssProps`, it warns for getting (and setting) of the `cssProps` property of `jQuery`.
